### ISSUE TITLE
Connection incredibly slow 

### ISSUE THREAD

**Issue by vintagexav** [robconery/massive-js#43]
For a <100MB db, with <100 tables and approx. 1000 stored procedures (procedures are mostly coming from postgis, geometry, and other default functions, only 7% of the procedures are custom):

loadTables: average `750ms`
loadFunctions: average `15000ms`

Connection string used: `postgres://user:pass@host:port/database?ssl=true`

massive.connect is being called before each request, passing that string as connection string. 
Therefore, `massive.loadTables` (`index.js`) is being called every time as well as `massive.loadFunctions`

This is performed from localhost, measured with `console.time` and `console.timeEnd`, and using pg.connect is almost instantaneous. 
Am I missing something? 

Extensions being used:
`postgis_topology, postgis, plv8, plpgsql, topology, fuzzystrmatch`

**Comment by xivSolutions** [comment]
Interesting. Possibly we should be checking to see if these items are already loaded during the call to connect. Seems like since tables, queries (stored in the local directory as .sql files), and functions all represent structural items unlikely to be modified during production use by an application, this should be safe. 

Then just make sure that any actions which DO cause structural changes to any of these things (and I don't think there are any at the moment) refresh the Db namespace when complete.

Could probably use some additional investigation as to the primary cause of the lag though. seems awfully slow for executing a couple `information_schema` queries, and iterating the result (even with some of the processing we have going on). Maybe the recursive walk through the file structure in `LoadQueries()`?

That's a lot of functions to load, but again, all that's really happening there is reading the function name, they aren't being executed or anything. 

Interesting to see what thinks, but I will do some investigating tomorrow too. Meantime, any additional info you find will be most helpful. 

@robconery  - Time maybe to set up a larger-scale perf test/demo against a production-size dataset, with a meaningful number of tables/functions, and a healthy dose of query files in the project directory?

**Comment by vintagexav** [comment]
Additional info: 

`node massive.js -c postgres://user:pass@host:port/database?ssl=true`

outputs:

```
runner
connect null
query null
loaded tables
runner
connect null
query null
loaded functions
```

Where, in runner.js I added 

- `console.log("runner");` after `DB.prototype.query = function () {`
- `console.log("connect",err);` after `assert.ok(err === null, err);`
- `console.log("query",err);` after `db.query(args.sql, args.params, function (err, result) {`

Where, in index.js I added

- `console.log("loaded tables");` after `massive.loadTables(function(err,db){`
- `console.log("loaded functions");` after `massive.loadFunctions(function(err,db){`

**Comment by xivSolutions** [comment]
I think the time you are seeing for `loadFunctions()` is also including the call to `LoadQueries().` Are you loading any query (.sql) files from the project directory as well?

**Comment by vintagexav** [comment]
In Navicat:

tables.sql:

```
select tc.table_schema as schema, tc.table_name as name, kc.column_name as pk
from 
    information_schema.table_constraints tc
    join information_schema.key_column_usage kc 
        on kc.table_name = tc.table_name
where 
    tc.constraint_type = 'PRIMARY KEY'
order by tc.table_schema,
         tc.table_name,
         kc.position_in_unique_constraint;
```

Outputs `47 records`

Takes, by itself, 0.25-0.30 secs - which is already unacceptable. 

and functions.sql:

```
SELECT routines.routine_schema as "schema", 
routines.routine_name as "name", routines.data_type,
(
  select count(1) from information_schema.parameters 
  where routines.specific_name=parameters.specific_name
    and parameter_mode = 'IN'
) as param_count
FROM information_schema.routines
WHERE routines."routine_schema" NOT IN('pg_catalog','information_schema')
AND routines.routine_name NOT LIKE 'pgp%'
order by routine_name
```

Outputs `1245` records

Takes 8-9 secs

**Comment by vintagexav** [comment]
@xivSolutions 
`node massive.js -c postgres://user:pass@host:port/database?ssl=true` does not load anything, it just runs the connection only - I didn't modify anything there.

**Comment by vintagexav** [comment]
This is accelerating things quite a bit (but still a unusable 1/2 sec response):

*functions.sql*:

```
SELECT routines.routine_schema as "schema", 
routines.routine_name as "name", routines.data_type
FROM information_schema.routines
WHERE routines."routine_schema" IN('public')
```

**Comment by vintagexav** [comment]
I suspect `postgis` to be the main problem, but still, this just raises a flag. 
Why load all tables and all functions every time whereas functions and tables could be passed as a String for free?

**Comment by robconery** [comment]
That's a lot of tables, and a *whole lot* of functions. My first reaction is that we should probably test to see which functions are from PostGis. My second reaction is that we should enable some kind of filter for table reads - meaning "Only load these" or "this pattern".

@vintagexav when you run the function sql do you see anything that would mark the function as Postgis? Also, would killing the function read help you in this case? 

Finally - a schema or two on your table would help as well - @xivsolutions we should have the ability to load only a particular schema.

**Comment by xivSolutions** [comment]
@robconery - love that idea (filterable on schema). I am also wondering what you think of doing the table/query/function load the first time you fire up the application, and from there checking to see if that stuff is already loaded before calling three backend queries every time a connection is made? 

As stated previously, this does not seem like the type of stuff that is likely to change often in a running production application. 

I would still want to see some better load times on some of this, and check to see if this is specific to the db/machine/configuration/whatever @vintagexav is working with, or in general is there some perf issue we should address. Gonna try playing with all of that today/tonight.

**Comment by robconery** [comment]
For some reason I didn't catch the nuance of the connection use here. That's a whole other issue:

> massive.connect is being called before each request

That shouldn't be the case, and it's not how Massive was designed (at least not how I use it). The idea is that you "bring Massive online" when the app starts. You capture the instance somewhere (maybe in your own module which is cached) and use it from there.

@vintagexav would this help your situation? @xivSolutions maybe we should make this use case clear in the README?

**Comment by xivSolutions** [comment]
Agreed on the connection usage, and example for the README. 

However, I am concerned about the 300-500 ms query times he indicates when running the tables query in Navicat - I can't replicate those results here on my machine with other sample data (using pgAdmin, and granted his database/table/function structure sounds like a unique snowflake). I get a more traditional 20-30 ms execution time. And the super-long 'load_functions' time, 1000-ish rows or no, seems suspicious as well. May need to pull down `postgis` and mess with it. 

I think we might want to set up a full-size/standard perf/demo db, and configure some samples. Maybe in a separate repo or something. Would be good, though, to be able to have some points of comparison when these things pop up.

**Comment by xivSolutions** [comment]
@vintagexav RE: Conncetion usage aside, I am still concerned with your query execution times in Navicat:

```
In Navicat:

tables.sql:

select tc.table_schema as schema, tc.table_name as name, kc.column_name as pk
from 
    information_schema.table_constraints tc
    join information_schema.key_column_usage kc 
        on kc.table_name = tc.table_name
where 
    tc.constraint_type = 'PRIMARY KEY'
order by tc.table_schema,
         tc.table_name,
         kc.position_in_unique_constraint;
Outputs 47 records
```

You say this is taking 250-300 ms?

If I run this very same query against the chinook database (30 records) I see execution time of ~30 ms using pgadmin. Not sure what could be different about the same query in your db, since it seems this should be unaffected (or marginally affected at best) by having `postgis` present (just a small handful of additional tables). Maybe a ton of key weird usage stuff to process in the join becuase of `postgis`? You are reporting an order of magnitude difference here, outside of the Massive context. 

If you are in fact getting execution times of 250-300 ms for this query in Navicat, then something is different, either in Navicat (which I doubt) or about your db. ??

**Comment by xivSolutions** [comment]
@vintagexav - Can you be bothereed to pull down [this (slightly modified)](https://github.com/xivSolutions/ChinookDb_Pg_Modified) implementation of the Chinook database, and running the tables query against it in Navicat and pgadmin? I am interested in comparing apples to apples here, and see if, in addition to the connection usage thing, there is something else going on?

**Comment by vintagexav** [comment]
I will try to look at it next week @xivSolutions. 
I'd be interested seeing what's the best way to call `massive.connect` only once @robconery

**Comment by robconery** [comment]
@vintagexav I'm happy to offer ideas - if you want to share some code I can try to help out. One way I've seen people do this is `app.set("db")` with Express (if that's what you're using) or, as mentioned, having a dedicated module that creates an instance and keeps it alive (though Massive does this already).

**Comment by xivSolutions** [comment]
I'm trying to come up with a very generalized example to add to the README. The challenge for me is, there are many different ways to approach the problem, depending on your needs and somewhat, how well you know Node - I'm still working on that second item ;-)

In the meantime @vintagexav, not knowing if you are using express, something else, or your own hoand-rolled routing, consider this brute-force/dumb-ish example to get started:

pretend you have a module you wrote that magically processed and routes requests:
```javascript
var assert = require("assert");

var Route = function(req, res, db) { 
  this.db = db;
  //...Magic happens....
  // ... and the request is properly routed to:
  this.get(req, res);
}

Route.prototype.get = function (req, res) { 
  // return all the artist records:
  this.db.artist.find(function (err, artists) { 
    assert(!err, err);
    res.writeHead(200, { 'Content-Type': 'application/json', "Access-Control-Allow-Origin":"*" });
    res.write(JSON.stringify(artists));
    res.end();
  });
};

module.exports = Route;
```
Then, consider an `index.js` file for your project like this:

```javascript
var massive = require("massive");
var path = require("path");
var Route = require("./lib/route");

var connectionString = "postgres://massive:password@localhost/chinook";
var scriptsDir = path.join(__dirname, "..", "db");

// stash the db here:
var db = {};

var configure = function(next) { 
  massive.connect({
  connectionString : connectionString,
  scripts : scriptsDir}, function(err, res) { 
    db = res;
  });
  // other configuration crap...
  next();
};

var handleRequests = function() { 
  var http = require('http');
  var port = 1337;
  http.createServer(function(req, res) {
    var route = new Route(req, res, db);
  }).listen(port);
};

// Start things up...
configure(handleRequests);
```
It's crude, and you wouldn't necessarily do things this way in a real application, but the concept is there anyway. @robconery if you see anything too dumb in the above, feel free to tune it up. 

I will say, coming up with an example that is clear is challenging, given the many ways one could approach this. I'm going to add something to the README that hopefully clearly illustrates the idea, and is equally clear that there are many ways to skin the cat.

**Comment by vintagexav** [comment]
`app.set('db',db_instance)` and `eq.app.get('db')` seems the most straightforward - definitely worth to go in the README. 

However:

- 10-20 seconds to init the db is a lot of downtime every time the app is rolled out. 
- Perfs are still just so bad. a `SELECT FROM WHERE id = 1` on a table with approx. 1k rows is <100ms using node's `pg.js` `connect` and `query` functions whereas it sums up above 500ms for Massive (and I am rounding in favor of Massive)

FYI (@xivSolutions) all my tests here and above are performed over network, not localhost.

**Comment by xivSolutions** [comment]
Ah.. Ok. Didn't want to assume you were using Express, and should have caught the network host in your previous comment. 

that IS some odd perf though...when you consider that with `run()` all that's really happening is Massive plucks out the args needed, and passes things on to pg `query()' anyway, then returns the result. All that expensive initialization is already done, so hard to see where the extra 400 + ms is coming from...

**Comment by vintagexav** [comment]
Gonna try to break it down.

**Comment by vintagexav** [comment]
Looks like pg.js is optimizing use of the connection / leverages cache.
The first request with pg is pretty much as slow as massive select. 
However the next ones are significantly faster with pg. 

*local srvr->real_world db: *

1st pg select: 500ms
1st massive select: 500ms
2nd pg select: 100ms
2nd massive select: still 500ms


*real srvr->real_world db: *

1st pg select: 25ms
1st massive select: 25ms
2nd pg select: 1-2ms
2nd massive select: still 25ms

Then I broke down each request.
I realized `pg-connect` is 0ms after the first request (then sometimes reaches back to similar level than `massive-connect`) whereas `massive-connect` is always as bad as 90% of the time needed to perform the select and never reaches the close to null time that `pg-connect` reaches.

**Comment by xivSolutions** [comment]
I wonder if this is related to [pg client or connection pooling](https://github.com/brianc/node-postgres/wiki/pg#method-connect)? Some where I think we're missing something silly here.

**Comment by xivSolutions** [comment]
I wonder if because we are passing the db instance in the call to `done()` (making it a "truthy" value) we are destroying the client with each call (per [this section of briancc's wiki entry](https://github.com/brianc/node-postgres/wiki/pg#parameters)), and failing to take advantage of the client pooling?

**Comment by robconery** [comment]
Let's keep this thread on track. @vintagexav the first thing that we need to establish is that *you* see this slow down (10-20s) for *your* situation (100+ tables, 1200+ functions). This is **not our targeted Use Case**. 

That said, it's something I very much want to accomodate. The queries as you see them, on their own, are not necessarily slow. What makes this whole thing slow is that Massive is trying to wrap close to 2000 objects and callbacks **which is utterly ridiculous**.

To keep us on track: you need to get this spin up time to come down, I need to see if we can help you with the way you're instantiating Massive. Both of the these things are doable and simple. There are 3 things to consider:

 - *can you use a schema* to namespace the things you want Massive to work with
 - *can we come up with a filter strategy* so that you can decide, on your own, which tables/functions you want spun up (if any)
 - *is Massive the right tool here*

The latter is a question we haven't brought up yet, but I'll be honest in saying that I didn't design it for your scenario (as valid as it is) and I'm not particularly inclined to add a bunch of code to accomodate for this scenario as I don't find it to be *typical*.

Before you get mad at me :):):) I think there's a good chance we can get some good middle ground that helps us all.

So - action items here are 1) help us come up with a filter strategy for you. We need to load tables/functions - what would work for you? and 2) would using a schema mitigate some of your issues?

I really mean it when I say *thank you* for your time. Please don't take any comment above as dismissive.

**Comment by xivSolutions** [comment]
I think that might have been it (or, part of the issue, anyway). Remove the `db` passed to the call to `done()` in `runner.js` and all the tests run over 2X faster (when using the `List` reporter, which gives individual test execution times). @robconery did I miss something here? I did a fix, and pushed before I saw your comment above, but the result was immediately apparent in test execution times.

**Comment by robconery** [comment]
As long as we're sure the connection is freed up - yay!

**Comment by xivSolutions** [comment]
According to brianc's wiki, calling `done()` returns the client to the pool, ready for another connection, but calling `done(db)` (or passing any "truthy" value) actually causes the client to be removed from the pool and and destroyed. 

@vintagexav - can you dig into `node_modules` (if you pulled massive down from npm - this fix isn't in there yet), do the same fix, and see if it at least corrects the consistency issues between pg and massive execution times for your project? there will likely still be long-ish load for your specific case, but interested to see the impact of the change here too. Then we can look at some of the things Rob suggests for filtering namespaces. 

I like the filtering idea in general anyway...

**Comment by robconery** [comment]
Radical - great catch John :).

**Comment by tracker1** [comment]
@robconery another suggestion would be to keep a cache of pools, and return the same pool for the same connectionstring/parameters...  That should reduce the chances of accidentally creating multiple pools that load in the same resources... I did this for `mssq-ng` which worked out pretty well.

**Comment by robconery** [comment]
Thanks @tracker1 - but the issue isn't connections (as far as I can tell) and pooling is handled by the driver, which apparently I had backwards (oops!).

Closing this issue for now; I think I'll open a few follow-ons to address large table/function DBs.
